diff a/src/hotspot/share/compiler/oopMap.hpp b/src/hotspot/share/compiler/oopMap.hpp
--- a/src/hotspot/share/compiler/oopMap.hpp
+++ b/src/hotspot/share/compiler/oopMap.hpp
@@ -197,15 +197,20 @@
 };
 
 
 class OopMapSet : public ResourceObj {
   friend class VMStructs;
+  friend class OopClosureWalker;
  private:
   GrowableArray<OopMap*> _list;
 
   void add(OopMap* value) { _list.append(value); }
 
+  static void stack_oop_do(oop *p, OopClosure* oop_fn,
+                      GrowableArray<oop> *stack_oops,
+                      intptr_t *stack_base, intptr_t *stack_top);
+
  public:
   OopMapSet();
 
   // returns the number of OopMaps in this OopMapSet
   int size() const            { return _list.length(); }
