diff a/src/hotspot/share/gc/shared/cardTable.hpp b/src/hotspot/share/gc/shared/cardTable.hpp
--- a/src/hotspot/share/gc/shared/cardTable.hpp
+++ b/src/hotspot/share/gc/shared/cardTable.hpp
@@ -246,10 +246,12 @@
   // This would be the 0th element of _byte_map, if the heap started at 0x0.
   // But since the heap starts at some higher address, this points to somewhere
   // before the beginning of the actual _byte_map.
   CardValue* byte_map_base() const { return _byte_map_base; }
   bool scanned_concurrently() const { return _scanned_concurrently; }
+  size_t byte_map_top_offset() const { return uintptr_t(_whole_heap.end()); }
+  size_t byte_map_bottom_offset() const { return uintptr_t(_whole_heap.start()); }
 
   virtual bool is_in_young(oop obj) const = 0;
 
   // Print a description of the memory for the card table
   virtual void print_on(outputStream* st) const;
